tools/sgen/sgen-grep-binprot.c

```diff
@@ -71,29 +71,52 @@ read_stream (EntryStream *stream, void *out, size_t size)
 }
 
 static int
-read_entry (EntryStream *stream, void *data)
+read_entry (EntryStream *stream, void *data, const char **name)
 {
 	unsigned char type;
 	ssize_t size;
 
-	if (read_stream (stream, &type, 1) <= 0)
+	if (read_stream (stream, &type, 1) <= 0) {
+		*name = NULL;
 		return SGEN_PROTOCOL_EOF;
+	}
 	switch (TYPE (type)) {
 
 #define BEGIN_PROTOCOL_ENTRY0(method) \
-	case PROTOCOL_ID(method): size = 0; break;
+	case PROTOCOL_ID(method): \
+		size = 0; \
+		*name = #method + strlen ("binary_protocol_"); \
```

Comment:
I don’t like the repeated `strlen ("binary_protocol_")` everywhere, but I was too lazy to refactor it.